perfmetrics/scripts/bigquery/bigquery.py

```diff
@@ -175,10 +218,96 @@ def setup_dataset_and_tables(self):
+    """.format(self.project_id, self.dataset_id, ExperimentsGCSFuseBQ.CONFIGURATION_TABLE_ID, gcsfuse_flags, branch, config_name)
+
+    job = self._execute_query_and_check_for_error(query_check_config_exists)
+    result_count = job.result().total_rows
+
+    # If more than 1 result -> duplicate experiment configuration present -> throw error
+    if result_count > 1:
+      raise Exception("Duplicate experiment configurations exist. Data corrupted")
+
+    # If result empty, then experiment configuration not present -> insert new experiment configuration -> return configuration ID
+    elif result_count == 0:
+      table_ref = self.dataset_ref.table(ExperimentsGCSFuseBQ.CONFIGURATION_TABLE_ID)
+      table = self.client.get_table(table_ref)
+      uuid_str = str(uuid.uuid4())
+      rows_to_insert = [(uuid_str, config_name, gcsfuse_flags, branch, end_date)]
+      result = self.client.insert_rows(table, rows_to_insert)
+      self._validate_result(result)
+      return uuid_str
+
+    # If exactly one result -> update end date -> return configuration ID
+    else:
+      config_id = job.to_dataframe().iloc[0]['configuration_id']
+      query_update_end_date = """
+        UPDATE `{}.{}.{}`
+        SET end_date = '{}'
+        WHERE configuration_id = '{}'
+        """.format(self.project_id, self.dataset_id, ExperimentsGCSFuseBQ.CONFIGURATION_TABLE_ID, end_date, config_id)
+      self._execute_query_and_check_for_error(query_update_end_date)
+      return config_id
+
+  def upload_metrics_to_table(self, table_name, config_id, start_time_build, metrics_data):
+
+    """Uploads metrics_data to the table corresponding to 'table_name'.
+
+    Args:
+      table_name (str): Table name corresponding to the table id to which results are being uploaded
+                        It can have the following values:
+                        'fio': For uploading to read_write_fio_metrics table
+                        'vm': For uploading to read_write_vm_metrics table
+                        'list': For uploading to list_metrics table
+      config_id (str): config_id of the experiment for which results are being uploaded
+      start_time_build (timestamp): Start time of the build
+      metrics_data (list): A 2D list containing the experiment results
+    """
+
+    # Check if the configuration ID of the experiment is valid
+    config_valid = self._check_if_config_valid(config_id)
+
+    if not config_valid:
+      raise Exception("Invalid configuration ID")
 
+    if table_name not in ExperimentsGCSFuseBQ.table_dict:
+      raise ValueError("Invalid table name passed to bigquery module")
 
+    table_id = ExperimentsGCSFuseBQ.table_dict[table_name]
+    table_ref = self.dataset_ref.table(table_id)
+    table = self.client.get_table(table_ref)
 
+    for row in metrics_data:
+      row_to_insert = [
+          (config_id, start_time_build) + tuple(row)
+      ]
+      result = self.client.insert_rows(table, row_to_insert)
```

Comment:
1. What happens if the insert_rows func fail after writing nth row ?
2. It looks like `insert_rows` function can insert more than one row at a time. Is there any reason why we are inserting the rows one by one ?

Reply:
I'll modify the code to insert all rows in a single API call.
insert_rows processes each row insertion independently, so if a single insertion fails, the previous insertions will still be there. I will add a check before inserting rows so that a row that has already been inserted does not get inserted again in case we try to insert rows again after insertion of the nth row fails.